scheduler/generic_sched.go

```diff
@@ -471,17 +428,30 @@ func (s *GenericScheduler) computePlacements(destructive, place []placementResul
 			// stop the allocation before trying to find a replacement because this
 			// frees the resources currently used by the previous allocation.
 			stopPrevAlloc, stopPrevAllocDesc := missing.StopPreviousAlloc()
+			prevAllocation := missing.PreviousAllocation()
 			if stopPrevAlloc {
-				s.plan.AppendUpdate(missing.PreviousAllocation(), structs.AllocDesiredStatusStop, stopPrevAllocDesc, "")
+				s.plan.AppendUpdate(prevAllocation, structs.AllocDesiredStatusStop, stopPrevAllocDesc, "")
+			}
+
+			// Compute penalty nodes for rescheduled allocs
+			selectOptions := &SelectOptions{}
```

Comment:
I would think about making a separate method: `getSelectOptions(...) (*SelectOptions, error)`. You could capture all this added logic and the preferred node in it.